omniledger/contracts/coins.go

```diff
@@ -0,0 +1,124 @@
+package contracts
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+
+	"github.com/dedis/cothority/omniledger/service"
+	"github.com/dedis/onet/log"
+)
+
+// ContractCoinID denotes a contract that can store and transfer coins.
+var ContractCoinID = "coin"
+
+var olCoin = service.NewInstanceID([]byte("olCoin"))
+
+// ContractCoin is a coin implementation that holds one instance per coin.
+// If you spawn a new ContractCoin, it will create an account with a value
+// of 0 coins.
+// The following methods are available:
+//  - mint will add the number of coins in the argument "coins" to the
+//    current coin instance. The argument must be a 64-bit uint in LittleEndian
+//  - transfer will send the coins given in the argument "coins" to the
+//    instance given in the argument "destination". The "coins"-argument must
+//    be a 64-bit uint in LittleEndian. The "destination" must be a 64-bit
+//    instanceID
+//  - fetch takes "coins" out of the account and returns it as an output
+//    parameter for the next instruction to interpret.
+//  - store puts the coins given to the instance back into the account.
+// You can only delete a contractCoin instance if the account is empty.
+func ContractCoin(cdb service.CollectionView, inst service.Instruction, c []service.Coin) (sc []service.StateChange, cOut []service.Coin, err error) {
+	cOut = c
+	switch {
+	case inst.Spawn != nil:
```

Comment:
we should use the helper functions `inst.GetType() == SpawnType` instead, this will catch errors when more than 1 of spawn/invoke/delete is set.